kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,952 @@
+		serviceUpdates:  make(chan serviceObject, 1),
+		ipamBlocks:      make(map[string]model.KVPair),
+		ipPools:         make(map[string]api.IPPool),
+		serviceInformer: serviceInformer,
+		serviceLister:   v1lister.NewServiceLister(serviceInformer.GetIndexer()),
+		allocationTracker: allocationTracker{
+			servicesByIp: make(map[string]types.NamespacedName),
+			ipsByService: make(map[types.NamespacedName]map[string]bool),
+		},
+	}
+
+	lbc.RegisterWith(lbc.dataFeed)
+
+	serviceAdd := func(obj interface{}) {
+		if svc, ok := obj.(*v1.Service); ok {
+			handle, err := createHandle(svc)
+			if err != nil {
+				log.WithError(err).Error("Error creating handle for service")
+				return
+			}
+			lbc.serviceUpdates <- serviceObject{
+				handle:     handle,
+				updateType: serviceUpdateTypeADD,
+				namespacedName: types.NamespacedName{
+					Namespace: svc.Namespace,
+					Name:      svc.Name,
+				},
+			}
+		}
+	}
+
+	serviceUpdate := func(objNew interface{}, objOld interface{}) {
+		if svc, ok := objNew.(*v1.Service); ok {
+			handle, err := createHandle(svc)
+			if err != nil {
+				log.WithError(err).Error("Error creating handle for service")
+				return
+			}
+			lbc.serviceUpdates <- serviceObject{
+				handle:     handle,
+				updateType: serviceUpdateTypeUPDATE,
+				namespacedName: types.NamespacedName{
+					Namespace: svc.Namespace,
+					Name:      svc.Name,
+				},
+			}
+		}
+	}
+
+	serviceDelete := func(objNew interface{}) {
+		if svc, ok := objNew.(*v1.Service); ok {
+			handle, err := createHandle(svc)
+			if err != nil {
+				log.WithError(err).Error("Error creating handle for service")
+				return
+			}
+			lbc.serviceUpdates <- serviceObject{
+				handle:     handle,
+				updateType: serviceUpdateTypeDELETE,
+				namespacedName: types.NamespacedName{
+					Namespace: svc.Namespace,
+					Name:      svc.Name,
+				},
+			}
+		}
+	}
+
+	_, err := lbc.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    serviceAdd,
+		UpdateFunc: serviceUpdate,
+		DeleteFunc: serviceDelete,
+	})
+	if err != nil {
+		log.WithError(err).Fatal("Failed to add event handle for Service LoadBalancer")
+		return nil
+	}
+	return lbc
+}
+
+// Run starts the controller.
+func (c *loadBalancerController) Run(stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	log.Debug("Waiting to sync with Kubernetes API (Service)")
+	if !cache.WaitForNamedCacheSync("loadbalancer", stopCh, c.serviceInformer.HasSynced) {
+		log.Info("Failed to sync resources, received signal for controller to shut down.")
+		return
+	}
+
+	// Load services and assigned IPs into cache
+	svcObjs, err := c.getServiceObjectList()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to get service objects")
+	}
+	for _, svcObj := range svcObjs {
+		if IsCalicoManagedLoadBalancer(svcObj.service, c.cfg.AssignIPs) {
+			for _, ingress := range svcObj.service.Status.LoadBalancer.Ingress {
+				c.allocationTracker.assignAddress(svcObj.service, ingress.IP)
```

Comment:
We need to update the allocation tracker based on the IPAM DB, and not based on the LoadBalancer.Ingress fields. This is becuase we use the `allocationTracker` as a faster / cached lookup of the contents of the IPAM DB.

If we use the service Status, we are prone to bugs like these:

- Someone puts a bogus entry into the Status in order to confuse us (intentionally or accidentally - other controllers might mess with this field)
- We have IPs allocated, but the Service is deleted while the pod is down (we'll not realize). 

You can get this "for free" if you use the `onBlockUpdate` function to populate the allocationTracker.